lib/spack/spack/spec.py

```diff
@@ -4146,6 +4151,123 @@ def trim(self, dep_name):
                         new_dependencies.add(edge)
             spec._dependencies = new_dependencies
 
+    def _virtuals_provided(self, root):
+        """Return set of virtuals provided by self in the context of root"""
+        if root is self:
+            # Could be using any virtual the package can provide
+            return set(self.package.virtuals_provided)
+
+        virtuals = set()
+        hashes = [s.dag_hash() for s in root.traverse()]
+        in_edges = set(
+            [edge for edge in self.edges_from_dependents() if edge.parent.dag_hash() in hashes]
+        )
```

Comment:
I *think* I like this better:

```suggestion
        in_edges = [
            edge for edge in self.edges_from_dependents() 
            if id(root) in  set(id(p) for p in edge.parent.traverse(direction="parents"))
        ]
```

Whether it's more efficient kind of depends on the graph, but I think searching up typically involves less traversing than searching down.

Reply:
We can't do a simple upwards traversal, because specs can have multiple dependents and upwards traversal will capture ancestors outside the scope of `root`.

Reply:
Does it matter? Root ids will still be distinct.

I thought about this and I suppose it’s an argument that upward traversal isn’t actually bounded, or that this could easily go wrong later. I can buy that.